### ISSUE TITLE
Complex step derivatives

### ISSUE THREAD

**Issue by proteneer** [tensorflow/tensorflow#25218]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.12
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Is it possible to implement the complex-step derivative method as a way to compute gradients? This would presumably require a "fairly simple" change of traversing the Ops of the graph, copying each one, and modifying its args from Real to Complex numbers and doing a 1e-16 delta in imaginary space. It may allow for easier computation of full Jacobians and Hessians. 

**Will this change the current api? How?**
I'm not sure.

**Who will benefit with this feature?**
People who need full rank Jacobian/Hessians.

**Any Other info.**
Reference paper:

http://mdolab.engin.umich.edu/sites/default/files/Martins2003CSD.pdf

The trick is basically to convert real numbers into complex numbers, and doing a small delta in imaginary space. Since many of tensorflow's ops already naturally support complex numbers, this seems like something that's doable.

**Comment by skye** [comment]
I think this would be a considerable amount of work, since our gradient functions currently don't expect to work over complex numbers. Someone could possibly prototype this by writing a version that works with a small number of ops though and see how it looks.

**Comment by skye** [comment]
I'm marking this "contributions welcome" in case someone experienced with TF wants to investigate this. Please don't attempt to fully implement this without further discussion though :)

**Comment by musikisomorphie** [comment]
@skye, this looks like an interesting topic to me, can I give a try?

**Comment by skye** [comment]
Sure. Do you have a rough plan of attack? If possible, I would recommend prototyping this first to get an idea of how much work it would be and what the benefits are. If the prototype is promising, I think ideally this would go in https://github.com/tensorflow/addons. Just to set expectations, I'm skeptical that a project of this complexity will make sense to include in TF core, so keep that in mind if you decide to take it on.

**Comment by musikisomorphie** [comment]
**Implement analytic complex derivative instead of numeric complex derivative**

@skype, after reading the paper recommended by @proteneer, I am not sure complex step derivatives is necessary for tensorflow framework. Basically, it is a complex number version of numerical derivative approximation.

>   f'= (f(x + h) - f(h)) / h 

In backprop instead of using the above approximation,  we normally compute the exact analytic derivative of a real function.  Therefore, in consistence with the real number case, I suppose we should add the analytic derivate for complex valued function. I realize that tf already support  some complex valued ops,  I can start by extending some other math ops to complex case and modify the optimizer accordingly. For example, I can follow the implementation here:

> https://github.com/ChihebTrabelsi/deep_complex_networks    

What do you think?

**Comment by skye** [comment]
I'm not familiar with complex nets or derivatives, so I don't have a sense for how useful this will be. @proteneer can you explain more what the end goal is? You mentioned full rank Jacobians and Hessians, does @musikisomorphie 's suggestion also achieve this?

**Comment by proteneer** [comment]
This is far less complicated than what you guys think it is. It basically allows for an easy way to do forward-mode AD in an embarassingly parallel way:

``` python
import numpy as np
import tensorflow as tf
sess = tf.Session()

def cosine_squared(x):
    return tf.reduce_sum(tf.pow(tf.cos(x), 2))

inp = tf.convert_to_tensor([1.0, 2.0], dtype=np.float64)
result = cosine_squared(inp)
backprop = sess.run(tf.gradients(result, inp)[0])
print(backprop)

step = 1e-7
inp = np.array([1.0, 2.0], dtype=np.float64) + 1j * np.array([step, 0], dtype=np.float64)
result = sess.run(cosine_squared(inp))
dfdx = result.imag/step
print(dfdx)

step = 1e-7
inp = np.array([1.0, 2.0], dtype=np.float64) + 1j * np.array([0, step], dtype=np.float64)
result = sess.run(cosine_squared(inp))
dfdx = result.imag/step
print(dfdx)
```

prints:

```
[-0.90929743  0.7568025 ]
-0.9092974264015236
0.7568024946574002
```

Literally the only engineering that's needed is to wrap the dfdx's in a parallel_for. 

As such it becomes trivial to do full-rank jacobians (as opposed to the reduced variants) that we currently have. And doesn't require _any_ backpropagation via jacobian-vector-products. The complex step takes care of this for you, automagically.

**Comment by musikisomorphie** [comment]
@proteneer @skye, thank you for your response. 
You suggest computing complex-step derivative when traversing the Ops of the graph.
But what about the possibility that the numerical errors of this kind of derivative will propagate and accumulate through the graph, especially when the nn is very deep?  I guess that is the reason why we always compute the analytic derivative for each op. 
For sure for certain ops like Jacobian/Hessians I can implement a temporary backbp by computing their complex-step derivative.  Personally, I am not sure implementing complex-step derivative through an entire graph is a good idea. 
I have a deadline this week, I can start coding next week once we come up with a solution that all of us agree with.

**Comment by proteneer** [comment]
The whole point of this method is that it's *insensitive* to the choice of the step size and the typical numerical problems that plague finite-difference type of approaches. It should have the same error as prop and forward prop.
